homeassistant/components/esphome/__init__.py

```diff
@@ -199,93 +200,213 @@ async def on_login() -> None:
+            tries = self._tries
+        # If not first re-try, wait and print message
+        # Cap wait time at 1 minute. This is because while working on the
+        # device (e.g. soldering stuff), users don't want to have to wait
+        # a long time for their device to show up in HA again (this was
+        # mentioned a lot in early feedback)
+        tries = min(tries, 10)  # prevent OverflowError
+        wait_time = int(round(min(1.8 ** tries, 60.0)))
+        if tries == 1:
+            _LOGGER.info("Trying to reconnect to %s in the background", self._host)
+        _LOGGER.debug("Retrying %s in %d seconds", self._host, wait_time)
+        await asyncio.sleep(wait_time)
+        self._reconnect_event.set()
+
+    async def _try_connect(self):
+        """Try connecting to the API client."""
+        async with self._tries_lock:
+            tries = self._tries
+            self._tries += 1
 
         try:
-            await cli.connect(on_stop=try_connect, login=True)
+            await self._cli.connect(on_stop=self._on_disconnect, login=True)
         except APIConnectionError as error:
             level = logging.WARNING if tries == 0 else logging.DEBUG
             _LOGGER.log(
                 level,
                 "Can't connect to ESPHome API for %s (%s): %s",
-                entry.unique_id,
-                host,
+                self._entry.unique_id,
+                self._host,
                 error,
             )
             # Schedule re-connect in event loop in order not to delay HA
             # startup. First connect is scheduled in tracked tasks.
-            data.reconnect_task = hass.loop.create_task(
-                try_connect(tries + 1, is_disconnect=False)
-            )
+            self._hass.loop.create_task(self._wait_and_start_reconnect())
         else:
-            _LOGGER.info("Successfully connected to %s", host)
-            hass.async_create_task(on_login())
+            _LOGGER.info("Successfully connected to %s", self._host)
+            async with self._tries_lock:
+                self._tries = 0
+            async with self._connected_lock:
+                self._connected = True
+            self._hass.async_create_task(self._on_login())
+
+    async def _reconnect_loop(self):
+        while True:
+            # Wait and clear reconnection event
+            await self._reconnect_event.wait()
+            self._reconnect_event.clear()
+
+            # If in connected state, do not try to connect again.
+            async with self._connected_lock:
+                if self._connected:
+                    continue
+
+            # Check if the entry got removed or disabled, in which case we shouldn't reconnect
+            if self._entry.entry_id not in self._hass.data[DOMAIN]:
+                # When removing/disconnecting manually
+                return
+            device_registry = (
+                await self._hass.helpers.device_registry.async_get_registry()
+            )
```

Comment:
```suggestion
            device_registry = self._hass.helpers.device_registry.async_get()
```

`async_get_registry` is deprecated

Reply:
Hmm, when I add this change exactly like you suggested, this line (with `async_get()`) blocks forever.

No idea why this would be happening, I'm not getting wiser from [the source of the method](https://github.com/home-assistant/core/blob/365e8a74ee639381d8625daa2f501052c4b2a2e9/homeassistant/helpers/device_registry.py#L619-L638). Do you know what could be up here?

Reply:
Nevermind, figured it out. 

Apparently the `hass` parameter is automatically bound for both methods, but for the callback the binding process blocks forever. Sounds like a bug, but unfortunately don't have the time to investigate now.